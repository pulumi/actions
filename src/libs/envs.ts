import * as envalid from 'envalid';

export const environmentVariables = envalid.cleanEnv(process.env, {
  GITHUB_WORKSPACE: envalid.str(),
  PULUMI_ACCESS_TOKEN: envalid.str({
    default: '',
  }),
});

export type EnvironmentVariables = typeof environmentVariables;
