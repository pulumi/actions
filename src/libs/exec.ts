import * as aexec from '@actions/exec';

export interface ExecResult {
  success: boolean;
  stdout: string;
  stderr: string;
}

export const exec = async (
  command: string,
  args: string[] = [],
  silent?: boolean,
): Promise<ExecResult> => {
  const { exitCode, stdout, stderr } = await aexec.getExecOutput(command, args, {
    silent: silent,
    ignoreReturnCode: true,
  });

  return {
    success: exitCode === 0,
    stdout: stdout.trim(),
    stderr: stderr.trim(),
  };
};
