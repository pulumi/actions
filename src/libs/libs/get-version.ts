import got from 'got';
import * as rt from 'runtypes';
import { maxSatisfying } from 'semver';
import invariant from 'ts-invariant';

const VersionRt = rt.Record({
  version: rt.String,
  date: rt.String,
  downloads: rt.Record({
    'linux-x64': rt.String,
    'linux-arm64': rt.String,
    'darwin-x64': rt.String,
    'darwin-arm64': rt.String,
    'windows-x64': rt.String,
  }),
  checksums: rt.String,
  latest: rt.Boolean.optional(),
});
export type Version = rt.Static<typeof VersionRt>;
const VersionsRt = rt.Array(VersionRt);

export async function getVersionObject(range: string): Promise<Version> {
  const result = await got(
    'https://raw.githubusercontent.com/pulumi/docs/master/data/versions.json',
    { responseType: 'json' },
  );

  const versions = VersionsRt.check(result.body);

  if (range == 'latest') {
    const latest = versions.find((v) => v.latest);
    invariant(latest, 'expect a latest version to exists');
    return latest;
  }

  const resp = maxSatisfying(
    versions.map((v) => v.version),
    range,
  );

  if (resp === null) {
    throw new Error(
      'Could not find a version that satisfied the version range',
    );
  }

  const ver = versions.find((v) => v.version === resp);

  if (!ver) {
    throw new Error(
      'Could not find a version that satisfied the version range',
    );
  }

  return ver;
}
