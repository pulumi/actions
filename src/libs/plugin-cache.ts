import * as crypto from 'crypto';
import * as os from 'os';
import * as path from 'path';
import * as cache from '@actions/cache';
import * as core from '@actions/core';

export interface PluginCacheConfig {
  enabled: boolean;
  pluginsPath: string;
}

/**
 * Resolves the plugins path by expanding home directory if needed
 */
export function resolvePluginsPath(pluginsPath: string): string {
  if (pluginsPath.startsWith('~/')) {
    return path.join(os.homedir(), pluginsPath.slice(2));
  }
  return pluginsPath;
}

/**
 * Generates a cache key for plugins based on the runner OS and a content hash
 */
export function generateCacheKey(
  stackName: string,
  workDir: string,
  pulumiVersion: string,
): string {
  // Create a hash based on stack name, work directory, and Pulumi version
  // This ensures different projects/stacks can have different plugin requirements
  const content = `${stackName}-${workDir}-${pulumiVersion}`;
  const hash = crypto.createHash('sha256').update(content).digest('hex').slice(0, 8);
  
  return `pulumi-plugins-${process.platform}-${hash}`;
}

/**
 * Generates restore keys for cache fallbacks
 */
export function generateRestoreKeys(): string[] {
  return [
    `pulumi-plugins-${process.platform}-`,
    'pulumi-plugins-',
  ];
}

/**
 * Attempts to restore plugins from cache
 */
export async function restorePluginsCache(
  cacheConfig: PluginCacheConfig,
  cacheKey: string,
): Promise<string | undefined> {
  if (!cacheConfig.enabled) {
    core.debug('Plugin caching is disabled');
    return undefined;
  }

  const resolvedPath = resolvePluginsPath(cacheConfig.pluginsPath);
  const restoreKeys = generateRestoreKeys();

  core.info(`Attempting to restore plugins cache from: ${resolvedPath}`);
  core.debug(`Cache key: ${cacheKey}`);
  core.debug(`Restore keys: ${restoreKeys.join(', ')}`);

  try {
    const cacheHit = await cache.restoreCache([resolvedPath], cacheKey, restoreKeys);
    
    if (cacheHit) {
      core.info(`Plugins cache restored successfully. Cache hit: ${cacheHit}`);
      return cacheHit;
    } else {
      core.info('No plugins cache found. Plugins will be downloaded as needed.');
      return undefined;
    }
  } catch (error) {
    core.warning(`Failed to restore plugins cache: ${error.message}`);
    return undefined;
  }
}

/**
 * Saves plugins to cache after Pulumi operations
 */
export async function savePluginsCache(
  cacheConfig: PluginCacheConfig,
  cacheKey: string,
  restoredCacheKey?: string,
): Promise<void> {
  if (!cacheConfig.enabled) {
    core.debug('Plugin caching is disabled');
    return;
  }

  // Don't save cache if we had a perfect cache hit
  if (restoredCacheKey === cacheKey) {
    core.info('Cache hit was exact, skipping cache save');
    return;
  }

  const resolvedPath = resolvePluginsPath(cacheConfig.pluginsPath);

  core.info(`Attempting to save plugins cache to: ${resolvedPath}`);
  core.debug(`Cache key: ${cacheKey}`);

  try {
    const cacheId = await cache.saveCache([resolvedPath], cacheKey);
    core.info(`Plugins cache saved successfully. Cache ID: ${cacheId}`);
  } catch (error) {
    // Cache save failures shouldn't fail the entire action
    core.warning(`Failed to save plugins cache: ${error.message}`);
  }
}