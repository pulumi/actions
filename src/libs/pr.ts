import * as core from '@actions/core';
import { context, getOctokit } from '@actions/github';
import * as dedent from 'dedent';
import invariant from 'ts-invariant';
import { Config } from '../config';

export async function handlePullRequestMessage(
  config: Config,
  output: string,
): Promise<void> {
  const {
    githubToken,
    command,
    stackName,
    options: { editCommentOnPr },
  } = config

  const heading = `#### :tropical_drink: \`${command}\` on ${stackName}`;
  const summary = '<summary>Pulumi report</summary>';

  const rawBody = output.substring(0, 64_000)
  // a line break between heading and rawBody is needed
  // otherwise the backticks won't work as intended
  const body = dedent`
    ${heading}

    <details>
    ${summary}

    \`\`\`
    ${rawBody}
    \`\`\`
    ${rawBody.length === 64_000 ? '**Warn**: The output was too long and trimmed.' : ''}
    </details>
  `

  const { payload, repo } = context
  invariant(payload.pull_request, 'Missing pull request event data.')

  const octokit = getOctokit(githubToken)

  try {
    if (editCommentOnPr) {
      const { data: comments } = await octokit.rest.issues.listComments({
        ...repo,
        issue_number: payload.pull_request.number,
      })
      const comment = comments.find(comment => comment.body.startsWith(heading) && comment.body.includes(summary))

      // If comment exists, update it.
      if (comment) {
        await octokit.rest.issues.updateComment({
          ...repo,
          comment_id: comment.id,
          body,
        })
        return
      }
    }
  } catch {
    core.warning(
      'Not able to edit comment, defaulting to creating a new comment.',
    )
  }

  await octokit.rest.issues.createComment({
    ...repo,
    issue_number: payload.pull_request.number,
    body,
  })
}
