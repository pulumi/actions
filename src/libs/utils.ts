export function parseArray(input: string): string[] {
  return parseUndefined(input)
    ? input.split(/\r?\n/).reduce<string[]>(
      (acc, line) =>
        acc
          .concat(line.split(','))
          .filter(pat => pat)
          .map(pat => pat.trim()),
      [],
    )
    : undefined
}

export function parseUndefined(input: string): string | undefined {
  return input === undefined || input === '' ? undefined : input
}

export function parseBoolean(input: string): boolean | undefined {
  return parseUndefined(input) ? input === 'true' : undefined
}

export function parseNumber(input: string): number | undefined {
  return parseUndefined(input) ? Number(input) : undefined
}
