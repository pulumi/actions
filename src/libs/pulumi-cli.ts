import * as os from 'os'
import * as path from 'path'
import * as core from '@actions/core'
import * as io from '@actions/io'
import * as tc from '@actions/tool-cache'
import * as exec from './exec'
import { getVersionObject } from './libs/get-version'

export async function isAvailable(): Promise<boolean> {
  const res = await exec.exec(`pulumi`, [], true)
  return res.stderr != '' && !res.success ? false : res.success
}

export async function run(...args: string[]): Promise<void> {
  await exec.exec(`pulumi`, args, true)
}

export function getPlatform(): string | undefined {
  const platforms = {
    'linux-x64': 'linux-x64',
    'linux-arm64': 'linux-arm64',
    'darwin-x64': 'darwin-x64',
    'darwin-arm64': 'darwin-arm64',
    'win32-x64': 'windows-x64',
  }

  const runnerPlatform = os.platform()
  const runnerArch = os.arch()

  return platforms[`${runnerPlatform}-${runnerArch}`]
}

export async function downloadCli(range: string): Promise<void> {
  const platform = getPlatform()

  if (!platform) {
    throw new Error(
      'Unsupported operating system - Pulumi CLI is only released for Darwin (x64, arm64), Linux (x64, arm64) and Windows (x64)',
    )
  }

  core.info(`Configured range: ${range}`)

  const { version, downloads } = await getVersionObject(range)
  core.info(`Matched version: ${version}`)

  const destination = path.join(os.homedir(), '.pulumi')
  core.info(`Install destination is ${destination}`)

  await io
    .rmRF(path.join(destination, 'bin'))
    .catch()
    .then(() => {
      core.info(`Successfully deleted pre-existing ${path.join(destination, 'bin')}`)
    })

  const downloaded = await tc.downloadTool(downloads[platform])
  core.debug(`Successfully downloaded ${downloads[platform]}`)

  switch (platform) {
    case 'windows': {
      await tc.extractZip(downloaded, os.homedir())
      await io.mv(path.join(os.homedir(), 'Pulumi'), path.join(os.homedir(), '.pulumi'))
      break
    }
    default: {
      await io.mkdirP(destination)
      core.debug(`Successfully created ${destination}`)
      const extractedPath = await tc.extractTar(downloaded, destination)
      core.debug(`Successfully extracted ${downloaded} to ${extractedPath}`)
      const oldPath = path.join(destination, 'pulumi')
      const newPath = path.join(destination, 'bin')
      await io.mv(oldPath, newPath)
      core.debug(`Successfully renamed ${oldPath} to ${newPath}`)
      break
    }
  }

  const cachedPath = await tc.cacheDir(path.join(destination, 'bin'), 'pulumi', version)
  core.addPath(cachedPath)
}
