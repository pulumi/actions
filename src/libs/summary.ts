import * as core from '@actions/core';
import { Config } from '../config';
import { stripAnsiControlCodes } from './utils';

function trimOutputByBytes(
  message: string,
  maxSize: number,
  alwaysIncludeSummary: boolean,
): [string, boolean] {
  /**
   *  Trim message to maxSize in bytes
   *  message: string to trim
   *  maxSize: Maximum number of bytes of final message
   *  alwaysIncludeSummary: if true, trim message from front (if trimming is needed), otherwise from end
   *
   *  return message and information if message was trimmed
   */
  let trimmed = false;

  const messageSize = Buffer.byteLength(message, 'utf8');

  // Check if message exceeds max size
  if (messageSize > maxSize) {

    // Trim input message by number of exceeded bytes from front or back as configured
    const dif: number = messageSize - maxSize;

    if (alwaysIncludeSummary) {
      message = Buffer.from(message).subarray(dif, messageSize).toString()
    } else {
      message = Buffer.from(message).subarray(0, messageSize - dif).toString()
    }

    trimmed = true;
  }

  return [message, trimmed];
}

export async function handleSummaryMessage(
  config: Config,
  projectName: string,
  output: string,
): Promise<void> {
  const {
    stackName,
    alwaysIncludeSummary,
  } = config;

  // Remove ANSI symbols from output because they are not supported in GitHub step Summary
  output = stripAnsiControlCodes(output);

  // Replace the first leading space in each line with a non-breaking space character to preserve the formatting
  const regex_space = RegExp(`^[ ]`, 'gm');
  output = output.replace(regex_space, '&nbsp;');

  // GitHub limits step Summary to 1 MiB (1_048_576 bytes), use lower max to keep buffer for variable values
  const MAX_SUMMARY_SIZE_BYTES = 1_000_000;

  const [message, trimmed]: [string, boolean] = trimOutputByBytes(output, MAX_SUMMARY_SIZE_BYTES, alwaysIncludeSummary);

  let heading = `Pulumi ${projectName}/${stackName} results`;

  if (trimmed && alwaysIncludeSummary) {
    heading += ' :warning: **Warn**: The output was too long and trimmed from the front.';
  } else if (trimmed && !alwaysIncludeSummary) {
    heading += ' :warning: **Warn**: The output was too long and trimmed.';
  }

  await core.summary
    .addHeading(heading)
    .addCodeBlock(message, "diff")
    .write();
}
