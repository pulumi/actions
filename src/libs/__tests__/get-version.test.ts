import * as playback from 'jest-playback';
import { getVersionObject } from '../libs/get-version';
playback.setup(__dirname);

describe('get-version', () => {
  process.env.GITHUB_TOKEN = process.env.GITHUB_TOKEN || 'my-token';
  describe('range versions', () => {
    it.each([
      'latest',
      '^3',
      '3.*.*',
      '3.0.*',
      'v1.*.*',
      'v2.22.*',
      '2.17.1',
      'v2.17.2',
      '2.5.0',
    ] as const)('should match %s versions', async (ver) => {
      const v = await getVersionObject(ver);
      expect(v.version).toMatchSnapshot();
    });
  });
});
