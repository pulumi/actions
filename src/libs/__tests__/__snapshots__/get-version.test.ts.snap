// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-version range versions should match ^3 versions 1`] = `"v3.3.1"`;

exports[`get-version range versions should match 2.5.0 versions 1`] = `"v2.5.0"`;

exports[`get-version range versions should match 2.17.1 versions 1`] = `"v2.17.1"`;

exports[`get-version range versions should match 3.*.* versions 1`] = `"v3.3.1"`;

exports[`get-version range versions should match 3.0.* versions 1`] = `"v3.0.0"`;

exports[`get-version range versions should match latest versions 1`] = `"v3.3.1"`;

exports[`get-version range versions should match v1.*.* versions 1`] = `"v1.14.1"`;

exports[`get-version range versions should match v2.17.2 versions 1`] = `"v2.17.2"`;

exports[`get-version range versions should match v2.22.* versions 1`] = `"v2.22.0"`;
